File: applications/bank-account-sink/src/main/java/com/vmware/bank/account/sink/RabbitConfig.java
package com.vmware.bank.account.sink;

import com.rabbitmq.stream.OffsetSpecification;
import lombok.extern.slf4j.Slf4j;
import org.springframework.amqp.rabbit.connection.ConnectionNameStrategy;
import org.springframework.amqp.rabbit.connection.SimplePropertyValueConnectionNameStrategy;
import org.springframework.amqp.rabbit.listener.MessageListenerContainer;
import org.springframework.amqp.support.converter.Jackson2JsonMessageConverter;
import org.springframework.amqp.support.converter.MessageConverter;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.autoconfigure.amqp.EnvironmentBuilderCustomizer;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.cloud.stream.config.ListenerContainerCustomizer;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.rabbit.stream.listener.StreamListenerContainer;

@Configuration
@Slf4j
public class RabbitConfig {


    @Value("${spring.rabbitmq.username:guest}")
    private String username;

    @Value("${spring.rabbitmq.password:guest}")
    private String password;

    @Value("${spring.rabbitmq.host:localhost}")
    private String hostname;

    @Value("${spring.application.name}")
    private String applicationName;

    @Bean
    public ConnectionNameStrategy cns() {
        return new SimplePropertyValueConnectionNameStrategy("spring.application.name");
    }

    @Bean
    public EnvironmentBuilderCustomizer envConfig()
    {
        return builder -> builder.id(applicationName);
    }

    @Bean
    MessageConverter convert()
    {
        return new Jackson2JsonMessageConverter();
    }

    @ConditionalOnProperty(name = "rabbitmq.streaming.replay",havingValue = "true")
    @Bean
    ListenerContainerCustomizer<MessageListenerContainer> customizer() {
        return (cont, dest, group) -> {
            StreamListenerContainer container = (StreamListenerContainer) cont;
            container.setConsumerCustomizer((name, builder) -> {
                builder.subscriptionListener(context ->{
                    log.info("Replaying from the first record in the stream");
                    context.offsetSpecification(OffsetSpecification.first());
                });
            });
        };
    }

}

File: applications/bank-geode-sink/src/main/kotlin/com/vmware/financial/open/banking/bank/sink/RabbitConfig.kt
package com.vmware.financial.open.banking.account

import com.rabbitmq.stream.ConsumerBuilder
import com.rabbitmq.stream.OffsetSpecification
import org.springframework.amqp.rabbit.connection.ConnectionFactory
import org.springframework.amqp.rabbit.connection.ConnectionNameStrategy
import org.springframework.amqp.rabbit.connection.RabbitConnectionFactoryBean
import org.springframework.amqp.rabbit.connection.ThreadChannelConnectionFactory
import org.springframework.amqp.rabbit.listener.MessageListenerContainer
import org.springframework.amqp.support.converter.Jackson2JsonMessageConverter
import org.springframework.amqp.support.converter.MessageConverter
import org.springframework.beans.factory.annotation.Value
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.cloud.stream.config.ListenerContainerCustomizer
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Profile
import org.springframework.rabbit.stream.listener.StreamListenerContainer


/**
 * RabbitMQ Configuration
 * @author Gregory Green
 */
@Configuration
class RabbitConfig {

    @Value("\${spring.rabbitmq.username:guest}")
    private var username: String = "guest";

    @Value("\${spring.rabbitmq.password:guest}")
    private var password:  String = "guest";

    @Value("\${spring.rabbitmq.host:localhost}")
    private var hostname: String = "localhost";

    @Value("\${spring.application.name}")
    private var applicationName: String? = null


    @Bean
    fun connectionFactory( ) : ConnectionFactory
    {
        var factory = RabbitConnectionFactoryBean();
        factory.setHost(hostname);
        factory.setUsername(username);
        factory.setPassword(password);

        var tcf = ThreadChannelConnectionFactory(factory.rabbitConnectionFactory);
        tcf.setConnectionNameStrategy(ConnectionNameStrategy { applicationName!! })
        return tcf
    }

    @Bean
    fun convert() : MessageConverter
    {
        return Jackson2JsonMessageConverter();
    }


    @Bean
    @Profile("stream")
    @ConditionalOnProperty(name = ["rabbitmq.streaming.replay"],havingValue = "true")
    fun customizer(): ListenerContainerCustomizer<MessageListenerContainer>? {
        return ListenerContainerCustomizer { cont: MessageListenerContainer, dest: String?, group: String? ->
            val container = cont as StreamListenerContainer
            container.setConsumerCustomizer { name: String?, builder: ConsumerBuilder ->
                builder.offset(
                    OffsetSpecification.first()
                )
            }
        }
    }
}